src/main/java/chat/Chat.java

```diff
@@ -0,0 +1,132 @@
+        embeddedIpfs.start();
+        Thread shutdownHook = new Thread(() -> {
+            LOG.info("Stopping server...");
+            try {
+                embeddedIpfs.stop().join();
+            } catch (Exception ex) {
+                ex.printStackTrace();
+            }
+        });
+        Runtime.getRuntime().addShutdownHook(shutdownHook);
+
+        System.out.println("Enter PeerId of other node:");
+        Scanner in = new Scanner(System.in);
+        String peerIdStr = in.nextLine().trim();
+        if (peerIdStr.length() == 0) {
+            throw new IllegalArgumentException("Invalid PeerId");
+        }
+        Multihash targetNodeId = Multihash.fromBase58(peerIdStr);
+        PeerId targetPeerId = PeerId.fromBase58(targetNodeId.bareMultihash().toBase58());
+
+        Multiaddr[] addressesToDial = getAddresses(targetNodeId);
+
+        runChat(embeddedIpfs.node, embeddedIpfs.p2pHttp.get(), targetPeerId, addressesToDial);
+    }
+    private Multiaddr[] getAddresses(Multihash targetNodeId) {
+        AddressBook addressBook = embeddedIpfs.node.getAddressBook();
+        PeerId targetPeerId = PeerId.fromBase58(targetNodeId.bareMultihash().toBase58());
+        Optional<Multiaddr> targetAddressesOpt = addressBook.get(targetPeerId).join().stream().findFirst();
+        Multiaddr[] allAddresses = null;
+        if (targetAddressesOpt.isEmpty()) {
+            List<PeerAddresses> closestPeers = embeddedIpfs.dht.findClosestPeers(targetNodeId, 1, embeddedIpfs.node);
+            Optional<PeerAddresses> matching = closestPeers.stream().filter(p -> p.peerId.equals(targetNodeId)).findFirst();
+            if (matching.isEmpty()) {
+                throw new IllegalStateException("Target not found: " + targetNodeId);
+            }
+            PeerAddresses peer = matching.get();
+            allAddresses = peer.addresses.stream().map(a -> Multiaddr.fromString(a.toString())).toArray(Multiaddr[]::new);
+        }
+        Multiaddr[] addressesToDial = targetAddressesOpt.isPresent() ?
+                Arrays.asList(targetAddressesOpt.get()).toArray(Multiaddr[]::new)
+                : allAddresses;
+        return addressesToDial;
+    }
+    private void runChat(Host node, HttpProtocol.Binding p2pHttpBinding, PeerId targetPeerId, Multiaddr[] addressesToDial) {
+        System.out.println("Type message:");
+        Scanner in = new Scanner(System.in);
+        while (true) {
+            byte[] msg = in.nextLine().trim().getBytes();
+            FullHttpRequest httpRequest = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.POST, "/", Unpooled.copiedBuffer(msg));
+            httpRequest.headers().set(HttpHeaderNames.CONTENT_LENGTH, msg.length);
+            HttpProtocol.HttpController proxier = p2pHttpBinding.dial(node, targetPeerId, addressesToDial).getController().join();
+            proxier.send(httpRequest.retain()).join().release();
+        }
+    }
+
+    public static void main(String[] args) {
+        new Chat();
```

Comment:
Can we inline all this so there's just a single main method? No need to go through an instance. 
I think we should maybe upstream that getAddresses method as well, probably move it to EmbeddedIpfs class.